src/models/backbones/resnet.py

```diff
@@ -0,0 +1,342 @@
+            downsample: Downsample module.
+            attn_layer: Attention block.
+        """
+        super().__init__()
+        out_block_channels = out_channels * self.expansion
+
+        self.convbnact1 = ConvBnAct(in_channels, out_channels, kernel_size=1, padding=0, stride=1)
+        self.convbnact2 = ConvBnAct(out_channels, out_channels, kernel_size=3, padding=1, stride=stride)
+
+        self.conv = nn.Conv2d(out_channels, out_block_channels, kernel_size=1, stride=1, bias=False)
+        self.bn = nn.BatchNorm2d(out_block_channels)
+        self.act = nn.ReLU(inplace=True)
+        self.se = attn_layer(out_block_channels) if attn_layer is not None else None
+
+        self.downsample = downsample
+
+    def forward(self, x: torch.Tensor):
+        """Forward method."""
+        identity = x
+
+        out = self.convbnact1(x)
+        out = self.convbnact2(out)
+
+        out = self.conv(out)
+        out = self.bn(out)
+
+        if self.se is not None:
+            out = self.se(out)
+
+        if self.downsample is not None:
+            identity = self.downsample(x)
+
+        out += identity
+        out = self.act(out)
+
+        return out
+
+
+class ResNet(BaseModel):
+    """ResNet model."""
+
+    def __init__(self,
+                 block: Union[BasicBlock, Bottleneck],
+                 layers: List[int],
+                 in_chans: int = 3,
+                 block_args: Dict = None):
+        """Init ResNet.
+
+        Args:
+            block: Type of block.
+            layers: Number of layers.
+            in_chans: Input channels.
+            block_args: Arguments for block_args.
+        """
+        self.block_args = block_args or dict()
+        self.inplanes = 64
+        self.channels = [64, 128, 256, 512]
+        self.num_features = 512 * block.expansion
+
+        super().__init__()
+
+        self.convbnact = ConvBnAct(in_chans, self.channels[0], kernel_size=7, stride=2, padding=3)
+        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
+
+        self.layer1 = self.__make_layer(block, self.channels[0], layers[0], **self.block_args)
+        self.layer2 = self.__make_layer(block, self.channels[1], layers[1], stride=2, **self.block_args)
+        self.layer3 = self.__make_layer(block, self.channels[2], layers[2], stride=2, **self.block_args)
+        self.layer4 = self.__make_layer(block, self.channels[3], layers[3], stride=2, **self.block_args)
+
+        self.create_hooks(output_channels=self.channels * block.expansion,
+                          module_names=['layer1', 'layer2', 'layer3', 'layer4'],
+                          strides=[1, 2, 2, 2])
+
+    def __make_layer(self,
+                     block: Union[BasicBlock, Bottleneck],
+                     in_channel: int,
+                     blocks: int,
+                     stride: int = 1,
+                     **block_args) -> nn.Sequential:
+        """Create layer - abstraction for convenient work.
```

Comment:
Create Resnet module which contain downsample and one of BasicBlock or Bottleneck modules.